rules/opentitan.bzl

```diff
@@ -232,67 +326,209 @@ def opentitan_binary(
+    unsigned BIN file, a (signed and) unsigned flash VMEM file, and a (signed
+    and) unsigned scrambled flash VMEM file. Some of these output targets
+    perform a bazel transition to the RV32I toolchain to build the target under
+    the correct compiler.
+    Args:
+      @param name: The name of this rule.
+      @param platform: The target platform for the artifacts.
+      @param signing_keys: The signing keys for to sign each BIN file with.
+      @param per_device_deps: The deps for each of the hardware target.
+      @param output_signed: Whether or not to emit signed binary/VMEM files.
+      @param **kwargs: Arguments to forward to `opentitan_binary`.
+    Emits rules:
+      For each device in per_device_deps entry:
+        cc_binary              named: <name>_<device>
+        obj_transform          named: <name>_<device>_elf
+        obj_transform          named: <name>_<device>_bin
+        elf_to_dissassembly    named: <name>_<device>_dis
+        bin_to_flash_vmem      named: <name>_<device>_flash_vmem
+        scrambled_flash_vmem   named: <name>_<device>_scr_flash_vmem
+        optionally:
+          sign_bin             named: <name>_<device>_bin_signed_<key_name>
+          bin_to_flash_vmem    named: <name>_<device>_flash_vmem_signed_<key_name>
+          scrambled_flash_vmem named: <name>_<device>_scr_flash_vmem_signed_<key_name>
+      filegroup named: <name>
+          with all the generated rules
+    """
+
+    deps = kwargs.pop("deps", [])
+    targets = []
+    for (device, dev_deps) in per_device_deps.items():
+        devname = "{}_{}".format(name, device)
+
+        # Generate ELF, Binary, and Disassembly
+        targets.extend(opentitan_binary(
+            name = devname,
+            deps = deps + dev_deps,
+            **kwargs
+        ))
 
         if output_signed:
             for (key_name, key) in signing_keys.items():
+                # Sign the Binary.
                 targets.append(":" + devname + "_bin_signed_" + key_name)
-                sign_image(
+                sign_bin(
                     name = devname + "_bin_signed_" + key_name,
                     bin = devname + "_bin",
                     elf = devname + "_elf",
                     key = key,
                     key_name = key_name,
                 )
 
+                # Generate a VMEM64 from the signed binary.
+                targets.append(":" + devname + "_vmem64_signed_" + key_name)
+                bin_to_flash_vmem(
+                    name = devname + "_vmem64_signed_" + key_name,
+                    bin = devname + "_bin",
+                    word_size = 64,
+                )
+
+                # Scramble signed VMEM64.
+                targets.append(":" + devname + "_scr_vmem64_signed_" + key_name)
+                scramble_flash_vmem(
+                    name = devname + "_scr_vmem64_signed_" + key_name,
+                    vmem = devname + "_vmem64_signed_" + key_name,
```

Comment:
I would prefer that we used format() for these names.

Reply:
Fixed.